json_serializable/lib/src/utils.dart

```diff
@@ -38,5 +51,182 @@ String friendlyNameForElement(Element element) {
+  }
+
+  /// Returns the offset of given field/property in its source file – with a
+  /// preference for the getter if it's defined.
+  int _offsetFor(FieldElement e) {
+    if (e.getter != null && e.getter.nameOffset != e.nameOffset) {
+      assert(e.nameOffset == -1);
+      return e.getter.nameOffset;
+    }
+    return e.nameOffset;
+  }
+
+  return _offsetFor(a).compareTo(_offsetFor(b));
+}
+
+final _dartCoreObjectChecker = const TypeChecker.fromRuntime(Object);
+
+/// Returns the set of fields that are written to via factory.
+/// Includes final fields that are written to in the constructor parameter list
+/// Excludes remaining final fields, as they can't be set in the factory body
+/// and shouldn't generated with toJson
+Set<FieldElement> writeFactory(
+    StringBuffer buffer,
+    ClassElement classElement,
+    Map<String, FieldElement> fields,
+    String ignoreReason(String fieldName),
+    String deserializeForField(FieldElement field,
+        {ParameterElement ctorParam})) {
+  var fieldsSetByFactory = new Set<FieldElement>();
+  var className = classElement.displayName;
+  // Create the factory method
+
+  // Get the default constructor
+  var ctor = classElement.unnamedConstructor;
+  if (ctor == null) {
+    throw new UnsupportedError(
+        'The class `${classElement.name}` has no default constructor.');
+  }
+
+  var ctorArguments = <ParameterElement>[];
+  var ctorNamedArguments = <ParameterElement>[];
+
+  for (var arg in ctor.parameters) {
+    var field = fields[arg.name];
+
+    if (field == null) {
+      // ignore: deprecated_member_use
+      if (arg.parameterKind == ParameterKind.REQUIRED) {
+        var msg = 'Cannot populate the required constructor '
+            'argument: ${arg.displayName}.';
+
+        var additionalInfo = ignoreReason(arg.name);
+
+        if (additionalInfo != null) {
+          msg = '$msg $additionalInfo';
+        }
+
+        throw new UnsupportedError(msg);
+      }
+
+      continue;
+    }
+
+    // TODO: validate that the types match!
+    // ignore: deprecated_member_use
+    if (arg.parameterKind == ParameterKind.NAMED) {
+      ctorNamedArguments.add(arg);
+    } else {
+      ctorArguments.add(arg);
+    }
+    fieldsSetByFactory.add(field);
+  }
+
+  var undefinedArgs = [ctorArguments, ctorNamedArguments]
+      .expand((l) => l)
+      .where((pe) => pe.type.isUndefined)
+      .toList();
+  if (undefinedArgs.isNotEmpty) {
+    var description =
+        undefinedArgs.map((fe) => '`${fe.displayName}`').join(', ');
+
+    throw new InvalidGenerationSourceError(
+        'At least one constructor argument has an invalid type: $description.',
+        todo: 'Check names and imports.');
+  }
+
+  // find fields that aren't already set by the constructor and that aren't final
+  var remainingFieldsForFactoryBody = fields.values
+      .where((field) => !field.isFinal)
+      .toSet()
+      .difference(fieldsSetByFactory);
+
+  //
+  // Generate the static factory method
```

Comment:
This method is ~100 lines and big section comments like this make me think it could be broken up into different methods.

Reply:
I did a bit of cleanup. There is a lot of state in a few `var`s here. I'd have to create another class to ship them between functions, which seems like more noise than it's worth